### ISSUE TITLE
login using  "type": "authorized_user"

### ISSUE THREAD

**Issue by leo-baltus** [hashicorp/vault-plugin-auth-gcp#46]
I would like to login to vault using my user-credentials, validated by gcp, to give admin permissions to myself and my fellow team members without accessing the root token and without using a serviceaccount to keep the root-token secure, and to have auditing tell me who is doing what.

Preferrably I would like to use the gcloud credentials but that seems [not](https://github.com/hashicorp/vault/issues/1052#issuecomment-367449911) possible?

Maybe I am misunderstanding, the only other way to do this currently is to have GOOGLE_APPLICATION_CREDENTIALS point to a json, generated with `gcloud beta auth application-default login` wich results in a json containing `"type": "authorized_user"` instead of `  "type": "service_account"` which seems is [supported](https://github.com/hashicorp/vault-plugin-auth-gcp/blob/master/plugin/cli.go#L35).

Is there any other way to use `vault login -method=gcp` using user-credentials?

**Comment by emilymye** [comment]
@leo-baltus this backend works on the assumption that the user (or GCE instance) has an associated service account, as the JWT it authenticates to Vault with is signed by a Google-owned service account key. If you're using user credentials, it might make more sense to use the [JWT/OIDC backend](https://www.vaultproject.io/docs/auth/jwt.html) pointed to Google. https://developers.google.com/identity/protocols/OpenIDConnect should have instructions on how to set it up on Google side.

**Comment by leo-baltus** [comment]
Hi @emilymye, thanks for pointing this out.

Using Terraform with a non-serviceaccount (i.e. a user-account) using `GOOGLE_APPLICATION_CREDENTIALS` has worked for us. Do you have the impression that Terraform, given it is run by the user, should be using OIDC as well? 

BTW I think that using a 'user associated service account' would defeat the purpose of having a user-account in the first place, or is there some way of invalidating this serviceaccount when the user leaves?

**Comment by emilymye** [comment]
**tl;dr**: Vault expects authenticating entities for this auth backend to be service accounts. Unless you're already associating service accounts with users, it indeed makes less sense to use this backend for users :)

To ramble (a lot) more, there are two concepts that apply to Vault and auth backends:

1. **Tool credentials**: How you allow Vault or Terraform permissions to call GCP APIs i.e. if you were to look at GCP logs you'd see any call to GCP API by Vault/Terraform as "identity X of type `authorized_user`/`service_account`/etc made this call". 
2. **Vault Identity**: What identity Vault associates with an auth token. When you run `vault login`, you get a token back that you use in future requests to the Vault server (i.e. `curl -H "X-Vault-Token: $TOKEN" ...` Every time this token is used, Vault logs this as "identity X used this token to access secrets/foo". 

You can still use `authorized_user` credentials for your *tool credentials*. However, this backend expects a service account identity and thus a service-account JWT token. You need the tool credentials in this case when you are running `vault login` because you are actually using the client-side CLI helper that does the hard work of constructing a request that the Vault server expects.

This CLI helper sends a request to GCP to get a signed JWT token via the [IAM API](https://cloud.google.com/iam/reference/rest/v1/projects.serviceAccounts/signJwt), but this call to GCP must be authenticated, which is where your ADC (tool credentials) comes in. The helper uses these tool credentials to obtain some OAuth token (lets call it `$USER_TOKEN`) and calls out to GCP: 

```
curl -H "Authorization: Bearer ${USER_TOKEN}" \
  ... \
  https://iam.googleapis.com/v1/serviceAccounts/$SERVICE_ACCOUNT/signJwt
``` 
which returns some JWT (`$SA_JWT`) associated with a `$SERVICE_ACCOUNT`.

The Vault client helper then sends the request to the server, i.e. 

```
vault login -method=gcp jwt="${SA_JWT}" role=my-role
```

> BTW I think that using a 'user associated service account' would defeat the purpose of having a user-account in the first place, or is there some way of invalidating this serviceaccount when the user leaves?

Some orgs don't use Google user identities outside of GCP. They might give a user multiple service accounts per job/project/etc - i.e. they would rather manage GCP-specific access per service account then per user. 

There isn't a good way to automatically invalidate service accounts/keys in vanilla GCP (shameless plug for my [secrets backend](https://github.com/hashicorp/vault-plugin-secrets-gcp)) but I think the same argument about forgetting to revoke accounts after users leave could be applied to giving a user all the permissions they need and forgetting to revoke them while that user still exists, and it's easier to    automate create-destroy workflows for service accounts compared to company-lifetime user accounts.

I'm going to go ahead and close this out but let me know if you want more clarification!

**Comment by leo-baltus** [comment]
Let me clarify:

I run vault in a k8s cluster, its running under a dedicated service-account so to it has access to a KMS-key and a GS bucket. Much like [this](https://github.com/kelseyhightower/vault-on-google-kubernetes-engine).

Since we run everything on GCP I want my team members, including myself, to use GCP authentication against vault, to get a role/token. This token wil given us write access to secrets. Services probably are going to use approle.

Here's what I have done:

    vault auth enable gcp

    vault write auth/gcp/role/leo \
    type="iam" \
    project_id="mit-kubernetes-leo" \
    policies="dev,prod" \
    bound_service_accounts="leo@example.com,foobar@mit-kubernetes-leo.iam.gserviceaccount.com"

Using the service-account, this seems to work

    vault login -method=gcp role="leo" credentials=@service-account.json
    Success! You are now authenticated. The token information displayed below

However using my Human-acccount (leo@example.com)

    export CLOUDSDK_CONFIG=/tmp/foo
    gcloud auth application-default login

    vault login -method=gcp role="leo" credentials=@/tmp/foo/application_default_credentials.json
results in:

    Error authenticating: could not obtain service account from credentials (are you using Application Default Credentials?). You must provide a service account to authenticate as

Explicitly demanding a service account here seems artificial to me. What do you consider wrong about a non-service-account?

**Comment by ahume** [comment]
Realise this issue has been closed, but though it worth adding that we have the same use-case. We'd hoped that our already configured user accounts would be able to authenticate against the backend without having to create Service Accounts for hundreds of individuals.

@leo-baltus Did you make any progress on how to achieve this?

**Comment by emilymye** [comment]
@ahume @leo-baltus It's not that I consider using a user account to auth to Vault a bad/unnecessary use case, it's that _this functionality already exists_. You can enable both the GCP backend and the JWT backend, and then create two roles, one for each backend, that share the same policies:

```sh
vault auth enable gcp
vault write auth/gcp/role/leo \
   ...
   policies="dev,prod"
   bound_service_accounts="foobar@mit-kubernetes-leo.iam.gserviceaccount.com"

vault auth enable jwt 
vault write auth/jwt/role/leo \
   ...
   policies="dev,prod"
   user_claim="leo@google.com"

vault login -method=gcp role="leo"  jwt="..."
vault login -method=jwt role="leo" jwt="..."
```

The two login statements are the expected login calls for these backends, and **these JWTs are expected to have different formats**. When you call `vault login -method=gcp role="leo" credentials=@service-account.json`, you're ACTUALLY calling a CLI helper that generates this JWT for you.

That means, in order to support users in this backend, we need to add ability to:
1. Take in a JWT signed by a user key, detect this is not a service account, construct a **different** JWT request to send to the OIDC server to verify that the user exists, and then verify all the various role bindings (i.e. the user is X@example.com, the user is in group Y) (https://accounts.google.com/o/oauth2/v2/auth).
2. 


Now, I'm not saying we can't do this or that it's not a terrible idea. In fact, the more I actually look at the JWT/OIDC 
The reason we split the backend like this is because we wanted to eventually



(this is what the JWT/OIDC backend does) 

OR 

2. have to write a helper that takes in the credential file 

Each auth backends on login takes a role and JWT. This JWT describes a Google identity (a service account or user). Vault verifies the login identity and JWT against the correct third-party service. This repo/plugin focuses on verifying identities against the GCP IAM API, i.e. we call `https://cloud.google.com/iam/reference/rest/v1/projects.serviceAccounts.keys/get` specifically to verify that the service account and associated key are valid and exist. 

A human user (i.e. `authorized_user`) would need to be verified against the OIDC endpoint, i.e. https://accounts.google.com/o/oauth2/v2/auth.





**The confusion might be coming from the fact that, when you call  you're actually calling the [client-side helper that constructs the JWT for you](https://github.com/hashicorp/vault-plugin-auth-gcp/blob/a00186a9031fc8db194e9b6002eed76db0ded90c/plugin/cli.go#L101-L107) in the CLI tool**  .  





. It doesn't know what to do with the `authorized_users` type because the GCP APIs don't really know about service accounts. If we add `authorized_users` our backend then would need to know how to talk to OIDC/Google Auth endpoints (i.e. https://accounts.google.com/o/oauth2/v2/auth) to get information about users, which is exactly what the [JWT/OIDC backend](https://www.vaultproject.io/docs/auth/jwt.html) does. (i.e. @ahume use this backend).

So, what I'm saying is (for @ahume as well), this functionality already exists but you just to need to enable both backends and create a role under each that verifies a different thing; the service account for GCP (bound_service_accounts = foobar@mit-kubernetes-leo.iam.gserviceaccount.com) and the user account (user_claim = leo@example.com) for JWT/OIDC  


 you wanted to use the above role, what you instead would do is 
The way 

 We didn't add that functionality because it was be essentially adding the same verification of users that the
If you want to be able to give both service accounts and users access to "dev,prod" policies, what you actually do is enable both policies and add roles with those policies for either. 

  
@ahume Please see [JWT/OIDC backend](https://www.vaultproject.io/docs/auth/jwt.html)

**Comment by emilymye** [comment]
@ahume @leo-baltus Opening for more discussion, sorry for the confusion. 

I don't consider using a user account to auth to Vault a bad/unnecessary use case, it's just that _this functionality already exists_  in the [JWT/OIDC backend](https://www.vaultproject.io/docs/auth/jwt.html) and I didn't think it was a good idea to re-support it here. I'm also willing to bet that it would be very hard to convince the Vault team why we would add it, since when we were designing this backend with HashiCorp, they specifically said we shouldn't add Google Auth functionality or a separate backend that was going to do what the JWT/OIDC backend already does.

You can enable both the GCP backend and the JWT backend, and then create two roles, one for each backend, that share the same policies:

```sh
vault auth enable gcp
vault write auth/gcp/role/leo \
   ...
   policies="dev,prod"
   bound_service_accounts="foobar@mit-kubernetes-leo.iam.gserviceaccount.com"

vault auth enable jwt 
vault write auth/jwt/role/leo \
   ...
   policies="dev,prod"
   user_claim="leo@google.com"

vault login -method=gcp role="leo"  jwt="..."
vault login -method=jwt role="leo" jwt="..."
```

Note that this backend specifically expect signed JWTs as the actual input and proof-of-identity. When you call `vault login -method=gcp role="leo" credentials=@service-account.json`, you're calling a CLI helper that either uses the explicitly given private key in the credentials file. If you don't provide credentials or a JWT, we attempt to use Application Default Credentials to call to an [API method](https://cloud.google.com/iam/reference/rest/v1/projects.serviceAccounts/signJwt) **_for service accounts only_**. This is the real issue - we don't have access to a method like that for users.

**Comment by emilymye** [comment]
(separating comments to talk about what I can actually help with) 

**So, tl;dr from the comment above**: I don't think it makes sense to add this to the `gcp` backend. It would be essentially adding a large amount of new code that makes existing, generalized code in another backend just _slightly_ more specific.  I also think the official (Vault team) recommendation would be to use the JWT/OIDC backend.

**NOW, that being said** - @ahume @leo-baltus, I will say admit I've been mostly parroting this advice and I haven't looked too closely at the JWT/OIDC backend until very recently (i.e. I was trying to write examples for how you use these two backends together). The documentation is...sparse? Sparse. Very sparse. We can certainly add guides for Google Auth or a Google-specific helper to construct the required JWT from user credentials. I don't think those are going to be contributions to this repo but I can help with those. 

cc @sethvargo

**Comment by ahume** [comment]
@emilymye Thank-you for taking the time to explain the situation here.

It seems totally reasonable, and I agree documentation covering this might be a nice addition. It's certainly not clear to me at this point exactly what this is going to look like. I'm going to try and get this working over the next few days, so I sense a learning opportunity. :)

**Comment by leo-baltus** [comment]
@emilymye Thanks for clarifying about the duplication of functionality that would occur if user-accounts were  supported by the `gcp` backend. Though it seems counter intuitive to not use the `gcp` backend. 

@ahume all documentation I have seen about constructing a JWT seem to use serviceaccounts. So I got stuck.

**Comment by ahume** [comment]
@leo-baltus I think that's where I also am stuck. I have enabled and configured the JWT/OIDC backend with Google's OIDC discovery_url, but I do not know how to generate a JWT for a users credentials, or even if that's really the next step.

**Comment by leo-baltus** [comment]
@emilymye I think using JWT is not supported: https://github.com/google/oauth2l/issues/58

**Comment by emilymye** [comment]
@leo-baltus @ahume just saw this, will add a follow-up comment explaining how to do things. I'm also planning to write a codelab that we'll try to get released soon on how to use this backend with Google OIDC. I believe the JWT you need to obtain is the ID token (mentioned in [OIDC docs](https://developers.google.com/identity/protocols/OpenIDConnect#authenticatingtheuser)) but I'm looking for the least complicated way to obtain this token.

**Comment by emilymye** [comment]
@leo-baltus @ahume So after investigating further, I'm going to go ahead and retract my comments about the JWT auth method working as a replacement for Google. It seems that the format of the OIDC token accepted by that auth method is incompatible with the Google ID token. 

However, I'm also going to go ahead and close this issue - we for sure are not going to support this user authentication in this plugin. If we do, it will probably be in a separate google OIDC plugin, since the authentication/verification methods are so different. I'll ping if we decide in the future to start dev work on such a thing. There are also some plugins in development now that may or may not be accepted in the future, see this [OIDC backend](https://github.com/hashicorp/vault/issues/5305) (which would use browser login). I'll update this issue if we decide to contribute a plugin, depending on how future work pans out.




we won't plan to support this
